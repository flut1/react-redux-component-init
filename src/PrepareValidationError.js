import ExtendableError from './utils/ExtendableError';

class PrepareValidationError extends ExtendableError {}

export default PrepareValidationError;
