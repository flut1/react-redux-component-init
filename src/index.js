export { default as withInitAction } from './withInitAction';
export { default as initReducer } from './reducer';

export { default as initComponent } from './actions/initComponent';
export { default as prepareComponent } from './actions/prepareComponent';
export { default as prepareComponents } from './actions/prepareComponents';
export { default as setInitMode } from './actions/setInitMode';

export { default as PrepareValidationError } from './PrepareValidationError';
export * from './initMode';
export * from './initSelfMode';
export * as actionTypes from './actions/actionTypes';
