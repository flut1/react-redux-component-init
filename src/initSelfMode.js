/**
 * Possible values for the `selfInit` option of `withInitAction()`
 */

/**
 * The component will render immediately, even if `initAction` is still pending.
 * It is recommended to use this option and render a loading indicator or placeholder content
 * until `initAction` is resolved. This will give the user immediate feedback that something is
 * being loaded. While the `initAction` is pending, an `isInitializing` prop will be passed to
 * the component.
 * @type {string}
 */
export const INIT_SELF_ASYNC = 'ASYNC';

/**
 * This will cause the component not to be mounted until the first initialization has completed.
 * The component will remain mounted during any reinitialization.
 * @type {string}
 */
export const INIT_SELF_BLOCKING = 'BLOCKING';

/**
 * The same behavior as `INIT_SELF_BLOCKING` but it will also unmount the component during
 * reinitialization.
 * @type {string}
 */
export const INIT_SELF_UNMOUNT = 'UNMOUNT';

/**
 * Only execute initAction on the server (initMode == MODE_PREPARE). Initialization will
 * be skipped on the client.
 * @type {string}
 */
export const INIT_SELF_NEVER = 'NEVER';
