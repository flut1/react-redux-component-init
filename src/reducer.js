import { createSelector } from 'reselect';

import createShallowArrayCompareSelector from './utils/createShallowArrayCompareSelector';
import { INIT_COMPONENT, SET_INIT_MODE } from './actions/actionTypes';
import extractValuesForProps from './utils/extractValuesForProps';
import { MODE_PREPARE } from './initMode';

/**
 * Reducer function that manages the state for `react-redux-component-init`. This reducer should
 * be included using Redux's `combineReducers()` under the `init` key in the root of
 * the store state.
 */
export default (state = {
  mode: MODE_PREPARE,
  prepared: {},
  selfInit: {},
}, action) => {
  switch (action.type) {
    case SET_INIT_MODE:
      return {
        ...state,
        mode: action.payload,
      };
    case INIT_COMPONENT:
      if (action.payload.isPrepare) {
        return {
          ...state,
          prepared: {
            ...state.prepared,
            [action.payload.prepareKey]: action.payload.complete,
          },
        };
      }
      return {
        ...state,
        selfInit: {
          ...state.selfInit,
          [action.payload.prepareKey]: action.payload.complete,
        },
      };
    default:
      return state;
  }
};

const createComponentInitValuesSelector = ({
  componentId, initProps, options: { getPrepareKey },
}) =>
  createShallowArrayCompareSelector(
    (state, props) => extractValuesForProps(props, initProps),
    initValues => ({
      prepareKey: getPrepareKey(componentId, initValues),
      initValues,
    }),
  );

export const createComponentInitStateSelector = initConfig => createSelector(
  createComponentInitValuesSelector(initConfig),
  state => state.selfInit,
  state => state.prepared,
  ({ prepareKey, initValues }, selfInit, prepared) => ({
    prepareKey,
    initValues,
    selfInitState: selfInit[prepareKey],
    isPrepared: !!prepared[prepareKey],
  }),
);
