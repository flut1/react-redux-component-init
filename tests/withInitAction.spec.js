import React, { Component } from 'react';

import withInitAction, { clearComponentIds } from '../src/withInitAction';
import { INIT_SELF_ASYNC, INIT_SELF_BLOCKING } from '../src/initSelfMode';

/* eslint-disable react/prefer-stateless-function, react/no-multi-comp */
describe('withInitAction', () => {
  describe('with a non-functional component', () => {
    describe('and no initProps or options', () => {
      clearComponentIds();
      class FooComponent extends Component {
        render() {
          return <noscript />;
        }
      }

      const WithInit = withInitAction(() => Promise.resolve())(FooComponent);
      it('has an initConfig', () => expect(typeof WithInit.initConfig).toBe('object'));
      it('has a string componentId', () => {
        expect(typeof WithInit.initConfig.componentId).toBe('string');
        expect(WithInit.initConfig.componentId).not.toBe('');
      });
      it('has an allowLazy option of false', () => expect(WithInit.initConfig.options.allowLazy).toBe(false));
      it('has an empty array for initProps', () => {
        expect(WithInit.initConfig.initProps).toEqual([]);
      });
      it('has a initSelf option of INIT_SELF_ASYNC', () => {
        expect(WithInit.initConfig.options.initSelf).toBe(INIT_SELF_ASYNC);
      });
      it('has a reinitialize option of true', () => {
        expect(WithInit.initConfig.options.reinitialize).toBe(true);
      });
    });

    it('should set the options on the initConfig object', () => {
      clearComponentIds();
      class FooComponent extends Component {
        render() {
          return <noscript />;
        }
      }

      const getInitStateDummy = state => state.foo;
      const onErrorDummy = e => e;

      const WithInit = withInitAction(
        () => Promise.resolve(),
        {
          reinitialize: false,
          initSelf: INIT_SELF_BLOCKING,
          allowLazy: true,
          getInitState: getInitStateDummy,
          onError: onErrorDummy,
        },
      )(FooComponent);

      it('sets the reinitialize option', () => {
        expect(WithInit.initConfig.reinitialize).toBe(false);
      });
      it('sets the initSelf option', () => {
        expect(WithInit.initConfig.initSelf).toBe(INIT_SELF_BLOCKING);
      });
      it('sets the allowLazy option', () => {
        expect(WithInit.initConfig.allowLazy).toBe(true);
      });
      it('sets the getInitState option', () => {
        expect(WithInit.initConfig.getInitState).toBe(getInitStateDummy);
      });
      it('sets the onError option', () => {
        expect(WithInit.initConfig.onError).toBe(onErrorDummy);
      });
    });
  });

  describe('with a functional component', () => {
    clearComponentIds();
    const FooComponent = () => <noscript />;

    const WithInit = withInitAction(
      () => Promise.resolve(),
    )(FooComponent);

    it('has a string componentId', () => {
      expect(typeof WithInit.initConfig.componentId).toBe('string');
      expect(WithInit.initConfig.componentId).not.toBe('');
    });
  });

  it('throws an error for components without a name or displayName', () => {
    clearComponentIds();
    const FooComponent = () => <noscript />;
    delete FooComponent.name;

    expect(() => {
      // eslint-disable-next-line no-unused-vars
      const WithInit = withInitAction(
        () => Promise.resolve(),
      )(FooComponent);
    }).toThrow();
  });

  it('set the initProps', () => {
    clearComponentIds();
    const FooComponent = () => <noscript />;

    const WithInit = withInitAction(
      ['a', 'b', 'c'],
      () => Promise.resolve(),
    )(FooComponent);
    expect(WithInit.initConfig.initProps).toEqual(['a', 'b', 'c']);
  });
});
