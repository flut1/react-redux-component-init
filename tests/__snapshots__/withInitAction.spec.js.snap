// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withInitAction with a component that has been prepared and has completed re-initializing does not set an { isInitializing: true } prop 1`] = `
<div>
  completed
</div>
`;

exports[`withInitAction with a component that has been prepared and is re-initializing sets an { isInitializing: true } prop 1`] = `
<div>
  initializing
</div>
`;

exports[`withInitAction with a component that has been prepared does not set an { isInitializing: true } prop 1`] = `
<div>
  completed
</div>
`;
